### PR TITLE
Feature proposal: Add Quantity.inverse/1 and new 1/unit type of unit

### PR DESCRIPTION
**Pull request by lasseebert**
This is a handy feature allowing to calculate the inverse of a quantity. (1 / quantity)

It also introduces a new type of unit, `{:div, nil, String.t()}`, which makes sense in many use-cases, e.g. for frequencies where `Hz` == `1/sec` or for intermediate calculations/

E.g. `~Q[1 1/sec]` is interpreted as `value: 1, unit: {:div, nil, sec}`

The `inverse/1` function is also handy when we already have a `:div` unit in which case, the unit is just turned upside down.

@skovmand What do you think?

### CODE REVIEW

**Comment by skovmand** [comment]
I think it's a good idea to support these too. Nice examples you have given :+1: 

The only thing I'm a bit sceptical about at first sight is the `nil` nominator. Could this possible clash with other uses that we might want later on?
In a way we already support these units in e.g. `sec^-1`? (Sorry, I don't know how to write superscript).. Anyway, the answer to this is of course that this won't work with the calculations unless we start recognizing superscript in the units, which we don't :-)...

**Comment by lasseebert** [comment]
@skovmand I think `{:div, nil, unit}` is pretty self-explaining, but you're right that `nil` in that context is kind of an anti pattern. Since the storage of the unit is internal (one would probably use `inverse` or `parse` with `1/unit` fo construction), we could easily introduce a new unit type, e.g. `{:inv, String.t}`.

What do you think?

**Comment by skovmand** [comment]
@lasseebert What when the denominator is removed by multiplication:

```
iex(1)> import Quantity.Sigils
Quantity.Sigils
iex(2)> frequency = ~Q[12 1/sec]
~Q[12 1/sec]
iex(3)> revolutions = ~Q[12 1/sec] |> Quantity.mult(~Q[12 sec])
~Q[144 ]
```

Should we handle that and return a scalar in this case?

**Comment by skovmand** [comment]
@lasseebert Some thoughts on this :thinking: .. Instead of introducing a special case for this unit, could we perhaps use `{:div, 1, unit}` instead? Wouldn't that make sense? Then we could remove this implementation:

```
def inverse(%Quantity{unit: {:div, 1, unit}} = quantity) do
  Quantity.new(Decimal.div(1, quantity.value), unit)
end
```

and instead just have this:

```
def inverse(%Quantity{unit: {:div, unit_a, unit_b}} = quantity) do
  Quantity.new(Decimal.div(1, quantity.value), {:div, unit_b, unit_a})
end
```

Except that it would output:

```
iex(6)> frequency = ~Q[10 1/sec] |> Quantity.inverse
~Q[0.1 sec/1]
```

However we might handle that general situation by introducing a function which changes `{:div, unit, 1}` to `unit`

**Comment by lasseebert** [comment]
Good point about `mult` with the inverse. I have also had other problems that are similar. E.g. if I have some conversion-quantities, used for converting from a set of units to another set of units.

E.g.

`~Q[1000 L/m³]`
`~Q[1000 kWh/MWh]`

But when something is already the right unit, the factor should be `1`, breaking the contract that this is a list of quantities.
This made me think that perhaps we should introduce unit-less quantities (that behave like decimals|). 

I like your idea about `1` being a valid unit, which we could special case in `inverse`, `div` and `mult`. And also add support for `~Q[18]` which is `value: ~d[18], unit: 1`.

**Comment by lasseebert** [comment]
@skovmand:

* Commit 819f966d107e2fed792889d3faf89f2709b115bf: Added `1` as a valid unit and also be able to shorten a lot more unit expressions on `mult` and `div`. This lead to a lot of special cases, and I'm not even sure I covered them all.
* Commit e39ec973c606d5a0802fd13f0f41ca1800e5ef56: Generalize `shorten_unit` and do not accept any resulting unit that can not be written on the form defined in the spec.

This leads to a few backwards incompatibilities: 

* `div/2` no longer returns Decimal, but always returns Quantity (but perhaps with a `1` unit)
* Calculations will now fail if resulting unit is too complex. E.g. `{:mult, "a", {:mult, "b", "c"}}` is no longer accepted. Anything that involves two or more `mult` and `div` in the unit does not compute.

File: lib/quantity/math.ex [hunk 1]
```diff
@@ -222,4 +222,56 @@ defmodule Quantity.Math do
...
+      {[a, b], []} -> {:mult, a, b}
+      {[a], [b]} -> {:div, a, b}
+      # Everything below here is not valid, but we try anyway
+      {as, []} -> Enum.reduce(as, &{:mult, &1, &2})
+      {[], bs} -> {:div, 1, Enum.reduce(bs, &{:mult, &1, &2})}
+      {as, bs} -> {:div, Enum.reduce(as, &{:mult, &1, &2}), Enum.reduce(bs, &{:mult, &1, &2})}
+    end
+  end
+
+  defp shorten({numerators, denominators}) do
```
Comment (skovmand) [inline lib/quantity/math.ex:245]: 1) Suggestion: Implement

```
defp shorten(numerators, []), do: {numerators, []}
defp shorten([], denominators), do: {[], denominators}
```

and implement the sorting as a separate function


2. Suggested comment for `shorten/2`:
```
# Remove common elements in numerator and denominator
```

File: lib/quantity/math.ex [hunk 2]
```diff
@@ -222,4 +222,56 @@ defmodule Quantity.Math do
   def round(quantity, decimal_count) do
     Quantity.new(Decimal.round(quantity.value, decimal_count, :half_up), quantity.unit)
   end
+
+  defp reduce_unit(unit) do
```
Comment (skovmand) [inline lib/quantity/math.ex:226]: Suggested comment: 

```
# reduce a unit to a standard form
# apply standard transformations, e.g. a/1 = a, etc.
# multiple-layer divisions will be reduced to a single fraction with numerators and denominators
```

It's actually pretty hard. Can it be used?

File: test/math_test.exs [hunk 3]
```diff
@@ -35,4 +35,66 @@ defmodule Quantity.MathTest do
...
+  describe "mult" do
+    test "mutliplying by quantity without unit" do
+      assert Quantity.mult(~Q[12 bananas], ~Q[3]) == ~Q[36 bananas]
+    end
+
+    test "mutliplying by quantity with 1/unit" do
+      assert Quantity.mult(~Q[12 bananas], ~Q[3 1/s]) == ~Q[36 bananas/s]
+    end
+
+    test "mutliplying by quantity with 1/unit that matches" do
```
[the remark this answers is not in the document]
Comment (lasseebert) [reply test/math_test.exs:85]: I think this was fixed in a later commit

[the remark this answers is not in the document]
Comment (lasseebert) [reply test/math_test.exs]: Ah was looking the wrong place. It is now fixed in a6c0892

File: lib/quantity/math.ex (hunk 1 above)
Comment (lasseebert) [reply lib/quantity/math.ex]: Changes in bc1785f

1. Agree that sorting should be it's own thing :+1: I don't agree to add extra function clauses for cases that are already covered in the main case. What is the reason for that?
2. :+1:

File: lib/quantity/math.ex (hunk 2 above)
Comment (lasseebert) [reply lib/quantity/math.ex]: :+1: Added in 82c6f4e

**Comment by lasseebert** [comment]
@skovmand: Ready for review :)

File: lib/quantity.ex [hunk 4]
```diff
@@ -68,21 +72,42 @@ defmodule Quantity do
...
+      parse_mult_unit(unit_string)
+    end
+  end
+
+  defp parse_mult_unit(unit_string) do
+    unit_string
+    |> String.split("*")
+    |> Enum.map(&parse_base_unit/1)
+    |> Enum.reduce(&{:mult, &1, &2})
+  end
```
Comment (skovmand) [inline lib/quantity.ex:106]: Minor detail: I think this is nice, but also somewhat implicit, because it actually also parses non-mult units (because `acc` is returned immediately from `Enum.reduce` if the list has only one element).

Perhaps rename it to `parse_mult_or_single_unit`?

File: lib/quantity.ex [hunk 5]
```diff
@@ -111,14 +136,18 @@ defmodule Quantity do
 
     unit_string =
       case quantity.unit do
-        {:div, u1, u2} -> "#{u1}/#{u2}"
-        {:mult, u1, u2} -> "#{u1}*#{u2}"
-        unit -> unit
+        1 -> ""
+        unit -> " #{unit_to_string(unit)}"
       end
```
Comment (skovmand) [inline lib/quantity.ex:141]: This is nice :+1:

File: lib/quantity.ex [hunk 6]
```diff
@@ -111,14 +136,18 @@ defmodule Quantity do
...
 
-    "#{decimal_string} #{unit_string}"
+    "#{decimal_string}#{unit_string}"
   end
 
+  defp unit_to_string(1), do: "1"
+  defp unit_to_string(unit) when is_binary(unit), do: unit
+  defp unit_to_string({:div, a, b}), do: "#{unit_to_string(a)}/#{unit_to_string(b)}"
+  defp unit_to_string({:mult, a, b}), do: "#{unit_to_string(a)}*#{unit_to_string(b)}"
```
Comment (skovmand) [inline lib/quantity.ex:149]: Nice :+1:

File: lib/quantity.ex [hunk 7]
```diff
@@ -285,4 +323,62 @@ defmodule Quantity do
       "~Q[#{@for.to_string(quantity)}]"
     end
   end
+
+  # Normalizes unit to a standard form:
+  # * Shorten unit as much as possible
+  # * At most one :div (with possibly many :mults on each side)
+  # * All :mult units are sorted
+  # * Extra 1-units are removed
+  defp normalize_unit(unit) do
```
Comment (skovmand) [inline lib/quantity.ex:332]: :tada: Really nice!

File: test/math_test.exs [hunk 8]
```diff
@@ -35,4 +35,90 @@ defmodule Quantity.MathTest do
...
+      assert Quantity.mult(~Q[12 s], ~Q[3 1/s]) == ~Q[36]
+      assert Quantity.mult(~Q[12 s*s], ~Q[3 1/s]) == ~Q[36 s]
+      assert Quantity.mult(~Q[12 s/banana], ~Q[3 1/s]) == ~Q[36 1/banana]
+    end
+
+    test "multiplying with inverse unit" do
+      assert Quantity.mult(~Q[12 s/banana], ~Q[3 banana/s]) == ~Q[36]
+    end
+
+    test "when unit does not produce anything valid, it still returns something" do
```
Comment (skovmand) [inline test/math_test.exs:95]: The title here states "anything valid", but now it is actually valid (outputting `~Q[34 apples*bananas/s*s]`).. Perhaps rename to "multiplies a complex unit" or something similar?

File: lib/quantity.ex (hunk 7 above)
Comment (skovmand) [reply lib/quantity.ex:332]: Niels-two-hours-ago is happy with this :smile:

Comment (lasseebert) [reply lib/quantity.ex:332]: :slightly_smiling_face: :+1:

File: test/math_test.exs (hunk 8 above)
Comment (lasseebert) [reply test/math_test.exs]: Ah yes, forgot to fix that test. Done in 69e2415

File: lib/quantity.ex (hunk 4 above)
Comment (lasseebert) [reply lib/quantity.ex:106]: I think, logically, it parses :mult units, which also includes the edge case of only one base unit. Since another function, `parse_base_unit/1`, handles the base case, I think `parse_mult_unit` is good.
Ok?

File: test/math_test.exs (hunk 8 above)
Comment (skovmand) [reply test/math_test.exs]: Nice :+1:

**Comment by skovmand** [comment]
@lasseebert :+1: for https://github.com/wise-home/quantity/pull/20/commits/574512cf684f739769183a5e1819d3554e389f3d, very nice

**Comment by lasseebert** [comment]
@skovmand Squashed to 5 commits. Will merge and publish new version :tada:
